third_party/2and3/redis/client.pyi

```diff
@@ -79,7 +79,7 @@ class Redis(Generic[_StrType]):
         encoding_errors: Text = ...,
         charset: Optional[Text] = ...,
         errors: Optional[Text] = ...,
-        decode_responses: Optional[bool] = ...,
+        decode_responses: Literal[True] = ...,
```

Comment:
This can't have a default, since that would mean that `Redis.from_url(url)` would be `Redis[str]`. Unfortunately, just removing the default won't work, because a non-default argument can't follow a default argument. The solution is a bit cumbersome: Duplicate the branch. See my commit for details.

This is also why mypy_primer flags problems in optuna.

Reply:
Out of curiosity, if you find the time, could explain this issue a bit further please? I don't understand why `Redis.from_url(url)` would be `Redis[str]` if the default value is `decode_responses=False`.

Reply:
@jherbel would you mind explaining why does `decode_responses` has to be a `Literal`?
what's wrong with a simple `bool` instead? you allow both values either way..

I'm having trouble during a transition from python2 to python3 (we try to keep both versions supported during the transition phase), where we would like to evaluate this value at runtime

this forces us to either duplicate code and perform an `if -> True, else -> False` or just use `# type: ignore` in both cases it feels wrong and I can't find a justification for it

Reply:
@MPalarya Your problem is probably best discussed at https://github.com/python/typing/discussions.

Reply:
@MPalarya In short, we need `Literal[True/False]` to enable `mypy` to properly distinguish between the different overload variants. We need different function signatures (aka different input types) in the different overload variants, so we cannot use `bool` but must use `Literal[True/False]` instead.

I consider this a bit of a hack, however, the problem of course originates in the redis package itself. I would say that using the _value_ of an input argument (`decode_responses=True/False`) to determine the output type is simply not good practice. And this is why we need this little hack.